test/Utilities/VariableTemplates/test_complex_models_gpu.jl

```diff
@@ -0,0 +1,110 @@
+using Test
+using StaticArrays
+using ClimateMachine.VariableTemplates
+using CUDA
+using KernelAbstractions
+using KernelAbstractions.Extras: @unroll
+CUDA.allowscalar(false)
+# CUDA.allowscalar(true) # for CPU tests
+
+number_states(m) = varsize(vars_state(m, Int))
+
+@kernel function mem_copy_kernel!(
+    m::AbstractModel,
+    dst::AbstractArray{FT, N},
+    src::AbstractArray{FT, N},
+) where {FT, N}
+    @uniform begin
+        ns = number_states(m)
+        vs = vars_state(m, FT)
+        local_src = MArray{Tuple{ns}, FT}(undef)
+        local_dst = MArray{Tuple{ns}, FT}(undef)
+    end
+    i = @index(Group, Linear)
+    @inbounds begin
+        @unroll for s in 1:ns
+            local_src[s] = src[s, i]
+        end
+        mem_copy!(m, Vars{vs}(local_dst), Vars{vs}(local_src))
+        @unroll for s in 1:ns
+            dst[s, i] = local_dst[s]
+        end
+    end
+end
+
+function mem_copy!(m::ScalarModel, dst::Vars, src::Vars)
+    dst.x = src.x
+end
+
+unroll_map(f, N::Int, args...) = unroll_map(f, Val(N), args...)
+@generated function unroll_map(f::F, ::Val{N}, args...) where {F, N}
+    quote
+        Base.@_inline_meta
+        Base.Cartesian.@nexprs $N i -> f(i, args...)
+    end
+end
+
+function mem_copy!(m::NTupleContainingModel{N}, dst::Vars, src::Vars) where {N}
+    dst.vector_model.x = src.vector_model.x
+    dst.scalar_model.x = src.scalar_model.x
+
+    up = ntuple(i -> src.ntuple_model[i].scalar_model.x, N)
+
+    unroll_map(N) do i
+        dst.ntuple_model[i].scalar_model.x = up[i]
+    end
```

Comment:
I ran into this with the remainder model, and @vchuravy gave me the magic sauce. `ntuple' with `Val` and `Base.@_inline_meta`
```suggestion
    @inbounds ntuple(Val(N)) do i
        Base.@_inline_meta
        dst.ntuple_model[i].scalar_model.x = up[i]
    end
```
which removes the need for the generated function.

Reply:
Oh wow! I thought we may have been missing magic sauce 😄. This works. In fact, after playing around with this, I found that simply
```julia
    ntuple(N) do i
        dst.ntuple_model[i].scalar_model.x = up[i]
    end
```
also works! Thanks @jkozdon! I'll remove the generated function 🚀

Reply:
Yeah, although I don't know why it wouldn't unroll automatically since `N` is passed as a type parameter. Unfortunately, while this is working fine for the toy example here, _none_ of these solutions are working for the actual use-case that this was all designed for. Branch [here](https://github.com/CliMA/ClimateMachine.jl/blob/ck/EDMFInSrcGPUUnroll), driver [here](https://github.com/CliMA/ClimateMachine.jl/blob/ck/EDMFInSrcGPUUnroll/test/Atmos/EDMF/bomex_edmf.jl), and observed failure [here](https://github.com/CliMA/ClimateMachine.jl/blob/ck/EDMFInSrcGPUUnroll/test/Atmos/EDMF/edmf_kernels.jl#L305). 😞 @vchuravy I'm probably going to need your help with this one!

Reply:
Update! Fortunately, there was an easy fix: the `unroll_map` ended up working, but needed a tweak:

```julia
@generated function unroll_map(f::F, ::Val{N}, args...) where {F, N}
    quote
        Base.@_inline_meta
        Base.Cartesian.@nexprs $N i -> f(Val(i), args...)              < ---------- need to wrap i with Val
    end
end
```
And we needed to modify the `getindex` accordingly to accept `Val{i}... where {i}`. So relieved!

Reply:
Can you use the `Val(i)` syntax with the `ntuple` syntax?

Reply:
Trying...

Reply:
No,
```
    @inbounds ntuple(Val(N_up)) do i
        Base.@_inline_meta
        up_a[Val(i)].buoyancy = eps(FT)
    end
```
fails. I guess that by the time that `Val` is called, it's too late and has already left the type domain.

Reply:
Bummer. Was just a thought.

I don't understand the whole GPU compile stack, so I would just be hacking at it like you are.

Reply:
Something else is going on. Now I'm getting a failure somewhere else (when passing `i` through functions), and I think that the compiler is somehow, sometimes, able to eliminate `Base.getindex(v::AbstractVars, i::Int)`, but sometimes not. Still debugging to figure out what an end-goal solution is. Updates to come.

Reply:
Okay, I figured out _what_ was triggering a dynamic runtime call, but I don't know exactly _why_. It's a bit difficult to explain, and there seem to be a lot of edge cases where certain things work in some circumstances and others do not. Ultimately, for the use-case, the `unroll_map` with `Val` seems needed. Plus a few more pieces. Going to expand on this.